oak_functions/loader/src/server.rs

```diff
@@ -87,69 +86,58 @@ impl Policy {
-    /// the response equal to `self.constant_response_size_bytes`
+    /// Creates an HTTP response, with status `OK`, the input `response` as its body.
     ///
-    /// The zeros are added as the header value for header `oak-policy-padding`.
-    /// Returns an error if the size of the input `response` is larger than or equal to
-    /// `self.constant_response_size_bytes`.
-    fn pad_response(&self, response: FunctionsResponse) -> anyhow::Result<Response<Body>> {
-        let response_size = estimate_size(&response)?;
-        let required_padding = self.constant_response_size_bytes as i64 - response_size;
-        if required_padding < 0 {
-            panic!(
-                "Something is very wrong. Only call `pad_response` with responses that require padding."
-            )
+    /// To conform to the security policy, the response is padded by adding a number of unspecified
+    /// characters, to make the total size of the response equal to
+    /// `self.constant_response_size_bytes`. The padding is added as the value of the
+    /// `oak-policy-padding` header. If the size of the input `response` is larger than
+    /// `self.constant_response_size_bytes` and the response is not already indicating policy
+    /// violation, then the `response` is discarded and an HTTP response indicating the policy
+    /// violation is returned. If `self.constant_response_size_bytes` is too small even for the
+    /// policy violation response, or the response already indicates a policy violation, an error is
+    /// returned.
+    fn create_http_response(&self, response: FunctionsResponse) -> anyhow::Result<Response<Body>> {
+        // deserialize response and get its size.
+        let mut encoded = vec![];
+        response
+            .encode(&mut encoded)
+            .context("couldn't encode response")?;
+        let padding_length = self.constant_response_size_bytes as i64 - encoded.len() as i64;
+        if padding_length < 0 {
+            match FunctionsStatusCode::from_i32(response.status) {
+                None => anyhow::bail!("invalid status code"),
+                Some(FunctionsStatusCode::PolicyViolation) => anyhow::bail!(
+                    "constant_response_size_bytes specified by the policy is too small."
+                ),
+                _ => self.create_http_response(FunctionsResponse {
+                    status: FunctionsStatusCode::PolicyViolation as i32,
+                    body: "Reason: the response is too long".as_bytes().to_vec(),
+                }),
+            }
         } else {
-            let padding_bytes = vec![b'0'; required_padding as usize];
+            let padding_bytes = vec![b'0'; padding_length as usize];
             http::response::Builder::new()
-                .status(response.status_code)
+                .status(StatusCode::OK)
                 .header(POLICY_PADDING_HEADER, std::str::from_utf8(&padding_bytes)?)
```

Comment:
BTW does this work correctly in case of zero bytes? I'm worried about whether there is an extra whitespace between the name and the value of the header that only gets added if the value is non-empty. Could you double check? (And also write it in a comment for future reference).

Reply:
Nice, thanks. BTW this may be a small unit test that proves in code what you just wrote here, so that it keeps being checked.

Reply:
`http::response` and `hyper` do not provide a utility for serializing the response into the final transmitted byte array. Moreover, the none of the client libraries seem to provide the raw response received as bytes. So, the test has to start a server and use curl as the client. Can you think of an easier way of testing this?

Actually, we could have done this more easily as part of a bash client for the hello-world example, since the size of the response can be controlled by the request. In general, it would have been easier to test policies with such an example. I wonder if I should add a simple `policy` example. WDYT?

Reply:
Maybe we should move away from using the HTTP headers as padding (in another PR), and instead have a type that we serialize and pad at the same time before handing it over to HTTP or gRPC or anything else that may handle it. The "contract" may be that the response is of a fixed size, but the expectation is that the receiver will encrypt it before sending it out. Maybe we can do something like defining a trait for serializing and padding (at the same time) a Request object, something like 

```rust
trait FixedSizeSerializer {
  fn serialize(v: Value, size: usize) -> Vec<u8>
}
```

Reply:
Is the output byte vector the binary protobuf encoding of the padded value? If it is not a valid protobuf message, then we have to define a protocol for serializing and deserializing. I think it is easier to just rely on protobuf.

Reply:
I think that trailing 0 bytes in a byte array would not be a problem for decoding protos. So, I think it could be done by just initialising a fixed length 0-initialised byte array and passing it the to `prost::Message::encode`. It should encode the proto at the start and leave the rest as 0s. This entire fixed length byte array could then be sent in the body, and the client should be able to decode it correctly.

Reply:
Thanks. I'll try that.

Reply:
I get the following error when decoding protobuf with trailing 0s:
```
DecodeError { description: "invalid tag value: 0", stack: [] }
```

Reply:
I think using `encode_length_delimited` and `decode_length_delimited` can fix this.

Reply:
That is a pity. I don't think length-delimited encoding is standardised yet, or compatible with all protobuf libraries. See e.g. https://seb-nyberg.medium.com/length-delimited-protobuf-streams-a39ebc4a4565

Reply:
We are already using length-delimited encoding for the lookup data. But that is not client-facing. 
The other alternative is to include a padding field in the response body. So, we'll have:
```proto
message Response {
  StatusCode status = 1;
  bytes body = 2;
  bytes padding = 3;
}
```
WDYT?